### ISSUE TITLE
iptables-legacy in firewall module should not be used directly

### ISSUE THREAD

**Issue by badfiles** [webmin/webmin#1603]
On my system (Ubuntu 20.04 with kernel 5.16)
I have only one working alternative -- /usr/sbin/iptables-nft
```
root@srv:/# update-alternatives --list iptables
/usr/sbin/iptables-legacy
/usr/sbin/iptables-nft
```
and set it manually via alternatives.

BUT firewall module sees available commands
```
iptables-legacy-restore
iptables-legacy-save
```
and uses save/restore directly (around alternatives)
```
sub missing_firewall_commands
{
local $c;
foreach $c ("ip${ipvx}tables", "ip${ipvx}tables-restore", "ip${ipvx}tables-save") {
	return $c if (!&has_command($c));
	}
return undef;
}

# iptables_restore()
# Activates the current firewall rules, and returns any error
sub iptables_restore
{
local $rcmd = &has_command("ip${ipvx}tables-legacy-restore") ||
	      "ip${ipvx}tables-restore";
local $out = &backquote_logged("cd / && $rcmd <$ipvx_save 2>&1");
return $? ? "<pre>$out</pre>" : undef;
}

# iptables_save()
# Saves the active firewall rules, and returns any error
sub iptables_save
{
local $scmd = &has_command("ip${ipvx}tables-legacy-save") ||
	      "ip${ipvx}tables-save";
local $out = &backquote_logged("$scmd >$ipvx_save 2>&1");
return $? ? "<pre>$out</pre>" : undef;
}
```
to fix this I had to rename the broken commands, I think this should be fixed

**Comment by jcameron** [comment]
Are you using the `-legacy` mode commands? If not, you can un-install them and Webmin will use `iptables-save` and `iptables-restore`

**Comment by badfiles** [comment]
I need legacy commands for ip6tables, 
iptables-save/restore are controlled by alternatives.
Why use legacy directly, if alternatives present?
On my system fail2ban uses `nft` via alternatives, `webmin-iptables.service` uses `ntf` via alternatives,
but webmin itself uses `legacy` directly, I get unexpected behavior.

**Comment by jcameron** [comment]
We use the `iptables-legacy` command directly because if `iptables` is pointing to `iptables-nft`, the rules file syntax is different.

**Comment by badfiles** [comment]
you should rework `webmin-iptables.service` then

**Comment by jcameron** [comment]
Is `webmin-iptables.service` running the wrong `iptables-restore` command?

**Comment by badfiles** [comment]
```
ExecStart=/bin/sh -c '/sbin/iptables-restore </etc/webmin/firewall/iptables.save'
```
I believe it should be `iptables-legacy-restore` according to this discussion

**Comment by badfiles** [comment]
even more, path '/sbin/iptables-restore' is not from my system
'</etc/webmin/firewall/iptables.save' this may not be changed via config, I believe that is hardcoded

**Comment by jcameron** [comment]
You're right, the init script should use the -legacy command .. we'll fix that.

**Comment by jcameron** [comment]
@PizzaProgram which Linux distribution are you running there?

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.

**Comment by PizzaProgram** [comment]
It is not tested completely at all.
I'm still waiting for some answers.

@iliaross Good, that You have re-opened! 👍

**Comment by jcameron** [comment]
Not sure if it will entirely solve your problems, but the next Webmin release will include a new module for managing NFtables

**Comment by iliaross** [comment]
I'd say a new Joe's `nftables` module should be used over legacy `iptables`.

For WireGuard, it's on my table to create a fully functional Webmin module. I recommend opening a new ticket in the issue tracker with a WireGuard module request and explaining how you use it and which features you’d like to have.

**Comment by iliaross** [comment]
Current `firewall/firewall-lib.pl` creates the Webmin iptables boot action via `iptables_restore_command()`, which uses the alternatives-managed `iptables-restore` or `ip6tables-restore` first and falls back to the legacy restore commands if needed. 

The matching save helper follows the same pattern.

Closing now.
